avocado/core/job.py

```diff
@@ -535,7 +546,8 @@ def run(self):
             self.exitcode |= exit_codes.AVOCADO_FAIL
             return self.exitcode
         finally:
-            self.job_pre_post_dispatcher.map_methods('post', self)
+            if self.job_pre_post_dispatcher is not None:
+                self.job_pre_post_dispatcher.map_methods('post', self)
```

Comment:
Yep, I like the structure, it's even better as it avoids running `post` without `pre`. The only detail I'm missing here is that usually pre and post are handled together as they are bounded, which can be seen from the need to modify `post` in order to coupe with the new structure. I guess the patch to treat `post` is later in the series, but if possible I'd prefer to squash it into this patch.